cpp/src/arrow/extension/variable_shape_tensor.cc

```diff
@@ -0,0 +1,341 @@
+    return Status::Invalid("Cannot convert non-fixed-width values to Tensor.");
+  }
+  if (data_array->null_count() > 0) {
+    return Status::Invalid("Cannot convert data with nulls values to Tensor.");
+  }
+
+  auto permutation = ext_type->permutation();
+  if (permutation.empty()) {
+    for (int64_t j = 0; j < static_cast<int64_t>(ext_type->ndim()); ++j) {
+      permutation.emplace_back(j);
+    }
+  }
+
+  std::vector<int64_t> shape;
+  for (int64_t j = 0; j < static_cast<int64_t>(ext_type->ndim()); ++j) {
+    ARROW_ASSIGN_OR_RAISE(const auto size, shape_array->GetScalar(j));
+    auto size_value = internal::checked_pointer_cast<Int32Scalar>(size)->value;
+    if (size_value < 0) {
+      return Status::Invalid("shape must have non-negative values");
+    }
+    shape.push_back(std::move(size_value));
+  }
+  internal::Permute<int64_t>(permutation, &shape);
+
+  std::vector<std::string> dim_names = ext_type->dim_names();
+  if (!dim_names.empty()) {
+    internal::Permute<std::string>(permutation, &dim_names);
+  }
+
+  std::vector<int64_t> strides;
+  ARROW_CHECK_OK(internal::ComputeStrides(value_type, shape, permutation, &strides));
+
+  const auto byte_width = value_type->byte_width();
+  const auto start_position = data_array->offset() * byte_width;
+  const auto size = std::accumulate(shape.begin(), shape.end(), static_cast<int64_t>(1),
+                                    std::multiplies<>());
+
+  ARROW_ASSIGN_OR_RAISE(
+      const auto buffer,
+      SliceBufferSafe(data_array->data()->buffers[1], start_position, size * byte_width));
+
+  return Tensor::Make(value_type, std::move(buffer), std::move(shape), std::move(strides),
+                      ext_type->dim_names());
+}
+
+Result<std::shared_ptr<DataType>> VariableShapeTensorType::Make(
+    const std::shared_ptr<DataType>& value_type, const int32_t ndim,
+    const std::vector<int64_t>& permutation, const std::vector<std::string>& dim_names,
+    const std::vector<std::optional<int64_t>>& uniform_shape) {
+  if (!dim_names.empty() && dim_names.size() != static_cast<size_t>(ndim)) {
+    return Status::Invalid("dim_names size must match ndim. Expected: ", ndim,
+                           " Got: ", dim_names.size());
+  }
+  if (!uniform_shape.empty() && uniform_shape.size() != static_cast<size_t>(ndim)) {
+    return Status::Invalid("uniform_shape size must match ndim. Expected: ", ndim,
+                           " Got: ", uniform_shape.size());
+  }
+  if (!uniform_shape.empty()) {
+    for (const auto& v : uniform_shape) {
+      if (v.has_value() && v.value() < 0) {
+        return Status::Invalid("uniform_shape must have non-negative values");
+      }
+    }
+  }
+  if (!permutation.empty()) {
+    if (permutation.size() != static_cast<size_t>(ndim)) {
+      return Status::Invalid("permutation size must match ndim. Expected: ", ndim,
+                             " Got: ", permutation.size());
+    }
+
+    std::vector<int64_t> sorted_permutation = permutation;
```

Comment:
Isn't there a utility function that checks a permutation's validity already? Why rewrite it here?

Reply:
Indeed, using it now. Still need to check `permutation.size() == ndim` as `IsPermutationValid` doesn't check for it.